content/en/blog/_posts/2022-10-27-live-and-let-live-with-kluctl-and-ssa.md

```diff
@@ -0,0 +1,174 @@
+
+## Deciding which conflicts to ignore and which to force-apply
+
+Kluctl has a few simple rules to figure out if a conflict should be ignored
+or force-applied.
+
+It first checks the field's actor (the field manager) against a list of known
+tools that are well known to be used to perform manual modifications. These
+are for example `kubectl` and `k9s`. Any modifications performed with these tools
+are considered "temporary" and will be overwritten by Kluctl.
+
+If the field manager is not known by Kluctl, it will check if a force-apply is
+requested for that field. A force-apply can be requested in different ways:
+
+1. By passing `--force-apply` to Kluctl. This will cause ALL fields to be force-applied on conflicts.
+2. By adding the [`kluctl.io/force-apply=true`](https://kluctl.io/docs/reference/deployments/annotations/all-resources/#kluctlioforce-apply) annotation to the object in question. This will cause all fields of that object to be force-applied on conflicts.
+3. By adding the [`kluctl.io/force-apply-field=my.json.path`](https://kluctl.io/docs/reference/deployments/annotations/all-resources/#kluctlioforce-apply-field) annotation to the object in question. This cause only fields matching the JSON path to be force-applied on conflicts.
+
+## Future work in regard to conflict resolution
+
+I plan to add more ways to control the behaviour of conflict resolution. For
+example:
+
+1. It will be possible to force-apply on field level via the CLI.
+2. It will be possible to force-apply on field level but only for specific field managers.
+3. It will be possible to modify the list of known actors for which fields are always force-applied.
+4. Exclude fields from force-apply even if the field manager is in the list of well known tools.
+5. If more is needed, you can [open an issue](https://github.com/kluctl/kluctl/issues).
+
+## DevOps vs. PipelineOps vs. GitOps vs. Controllers
+
+So how does server-side apply in Kluctl lead to "live and let live"?
+
+It allows the co-existence of classical DevOps, PipelineOps, GitOps and
+controllers. You can use a GitOps style deployment pipeline (check 
+[flux-kluctl-controller](https://github.com/kluctl/flux-kluctl-controller))
+while having the proper tools (e.g. conflict resolution via annotations) to
+act on expected or unexpected modifications to your objects from other
+actors. The same applies to classical pipeline based deployments
+(e.g. Github Actions or Gitlab CI).
+
+You can intervene with your admin permissions if required and run a `kubectl`
+command that will modify a field and prevent Kluctl from overwriting it. You'd
+just have to switch to a field-manager (e.g. "admin-override") that is not
+overwritten by Kluctl.
+
+If a controller is known to erroneously claim fields (the ECK operator does this
+to the nodeSets field for example), you can ensure that Kluctl always overwrites
+these fields to the original or a new value.
```

Comment:
To end the article, what are your 3 recommendations to people doing advanced server-side apply in _their_ controller?

Reply:
Hmm good idea to add this. I was not really happy with the abrupt ending to be honst.

Reply:
Added a "A few takeaways" chapter which contains a few recommendations.